src/moments.jl

```diff
@@ -73,32 +73,30 @@ end
 
 function trace_product(o1::OperatorTS2D, o2::OperatorTS2D; scale=0)
     checklength(o1, o2)
-    N = qubitlength(o1)
-    L1 = extent(o1)
-    L2 = N ÷ L1
-    (scale == 0) && (scale = 2.0^N)
+    L1, L2 = qubitsize(o1)
+    (scale == 0) && (scale = 2.0^(L1*L2))
     tr = zero(scalartype(o1))
 
     # ensure `@inbounds` is safe
     length(o1.strings) == length(o1.coeffs) || throw(DimensionMismatch("strings and coefficients must have the same length"))
     length(o2.strings) == length(o2.coeffs) || throw(DimensionMismatch("strings and coefficients must have the same length"))
 
-    @inbounds for i1 in eachindex(o1.strings)
-        p1, c1 = o1.strings[i1], o1.coeffs[i1]
-        for i2 in eachindex(o2.strings)
-            p2, c2 = o2.strings[i2], o2.coeffs[i2]
+    @inbounds for (p1, c1) in zip(o1.strings, o1.coeffs)
+        rep1 = representative(p1)
+        for (p2, c2) in zip(o2.strings, o2.coeffs)
+            rep2 = representative(p2)
             for i in 0:L1-1
                 for j in 0:L2-1
-                    p3 = rotate_lower(p2, i, j, L1)
-                    p, k = prod(p1, p3)
+                    shifted = shift(rep2, Val((L1, L2)), (i, j))
```

Comment:
Doesn't this lead to a dynamic dispatch, kind of defeating the purpose of using the `Val`s?
Could you check to see if replacing this with regular integers actually makes things slower?